web-api-shared/src/elastic.rs

```diff
@@ -0,0 +1,341 @@
+        &self,
+        method: Method,
+        segments: impl IntoIterator<Item = &'a str>,
+        query_parts: impl IntoIterator<Item = (&'a str, Option<&'a str>)>,
+    ) -> RequestBuilder {
+        let url = self.create_url(segments, query_parts);
+        let builder = self.client.request(method, url);
+        self.auth.apply_to(builder)
+    }
+
+    pub fn create_url<'a>(
+        &self,
+        segments: impl IntoIterator<Item = &'a str>,
+        query_parts: impl IntoIterator<Item = (&'a str, Option<&'a str>)>,
+    ) -> Url {
+        let mut url: Url = self.url_to_index.with_segments(segments).into();
+        let mut query_mut = url.query_pairs_mut();
+        for (key, value) in query_parts {
+            if let Some(value) = value {
+                query_mut.append_pair(key, value);
+            } else {
+                query_mut.append_key_only(key);
+            }
+        }
+        drop(query_mut);
+        url
+    }
+}
+
+#[derive(Debug, Serialize)]
+#[serde(rename_all = "lowercase")]
+pub enum BulkInstruction<'a, I> {
+    Index {
+        #[serde(rename = "_id")]
+        id: &'a I,
+    },
+    Delete {
+        #[serde(rename = "_id")]
+        id: &'a I,
+    },
+}
+
+#[derive(Debug, Deserialize)]
+pub struct BulkItemResponse<I> {
+    #[serde(rename = "_id")]
+    pub id: I,
+    pub status: u16,
+    #[serde(default)]
+    pub error: Value,
+}
+
+impl<I> BulkItemResponse<I> {
+    fn is_success_status(&self, allow_not_found: bool) -> bool {
+        StatusCode::from_u16(self.status)
+            .map(|status| {
+                (status == StatusCode::NOT_FOUND && allow_not_found) || status.is_success()
+            })
+            .unwrap_or_default()
+    }
+}
+
+#[derive(Debug, Deserialize)]
+pub struct BulkResponse<I> {
+    pub errors: bool,
+    pub items: Vec<HashMap<String, BulkItemResponse<I>>>,
+}
+
+impl<I> BulkResponse<I> {
+    pub fn failed_documents(self, operation: &'static str, allow_not_found: bool) -> Vec<I>
+    where
+        I: Display + Debug,
+    {
+        self.errors.then(|| {
+            self
+                .items
+                .into_iter()
+                .filter_map(|mut response| {
+                    if let Some(response) = response.remove(operation) {
+                        if !response.is_success_status(allow_not_found) {
+                            error!(
+                                document_id=%response.id,
+                                error=%response.error,
+                                "Elastic failed to {operation} document.",
+                            );
+                            return Some(response.id);
+                        }
+                    } else {
+                        error!("Bulk {operation} request contains non {operation} responses: {response:?}");
+                    }
+                    None
+                })
+                .collect()
+        }).unwrap_or_default()
+    }
+}
+
+impl Client {
```

Comment:
🔧 this has conflicts with the recently merged #937 since some new code is missing.